### DISCUSSION TITLE
Catch undeclared (unimported) absolute paths

### DISCUSSION THREAD

**Discussion by dcecile** [rust-lang/rust-clippy#12002] [Q&A] [answered]
I'd like to have all absolute module usages declared via imports. Are there any existing lints that can do this? I've tried `absolute_path` but it didn't work [1].

Suppose that this is my whole file:

```rs
pub fn get_mime_type(ext: &str) -> Option<mime::Mime> {
  if ext == "html" {
    Some(mime::TEXT_HTML)
  } else {
    None
  }
}
```

I want Clippy to complain until the absolute module is explicitly imported:

```rs
use mime;

pub fn get_mime_type(ext: &str) -> Option<mime::Mime> {
  if ext == "html" {
    Some(mime::TEXT_HTML)
  } else {
    None
  }
}
```

---

[1] [absolute_paths](https://rust-lang.github.io/rust-clippy/master/index.html#/absolute_paths) with `absolute-paths-max-segments = 1` finds the problem, but does not allow resolving it adding an import line.

Instead it requires all children to be imported individually (not something I want to force):

```rs
use mime::Mime;
use mime::TEXT_HTML;

pub fn get_mime_type(ext: &str) -> Option<Mime> {
  if ext == "html" {
    Some(TEXT_HTML)
  } else {
    None
  }
}
```

**Comment by flip1995** [comment]
I would consider the `use mime`+`mime::Mime` suggestion as introducing an unused import. I'm more surprised that `rustc` doesn't warn about this. Why do you want to import it explicitly?

**Reply by dcecile** [reply]
I like the convention (from other programming languages) of having all module references listed at the top of the file. This gives me a good way at a glance to get initial context for the code in the file.

I'm also trying to avoid the [Ruby on Rails autoloading](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html) pattern where "application classes and modules are just available everywhere without `require` calls".

**Reply by flip1995** [reply, marked as answer]
The Ruby on Rails thing is not a pattern that works in Rust. If you don't import something, you can only access it through the full path, which is `crate::[modules::]*Item`. 

Back in Rust 2015, if you want to use something from an extern crate `A`, you had to write `extern crate A;`. Rust 2018 got rid of this restriction and made it more convenient to use something from an extern crate. What you're asking is pretty much a lint that adds this back with `use` statements. 

I don't think this lint would get a lot of use. But I'd accept it as an `restriction` lint, if you were to open a PR for it :)

**Reply by dcecile** [reply]
Got it, thanks!

**Comment by x-at-01** [comment]
If you are looking for an automated tool to refactor existing absolute/qualified paths into top-level `use` statements across your codebase (to satisfy `clippy::absolute_paths`), take a look at [fixrs](https://github.com/webc-site/fixrs).

### Demo

**Before:**
```rust
fn handle_data() -> Result<(), std::io::Error> {
    let now = std::time::Instant::now();
    let code = std::io::ErrorKind::NotFound;
    Ok(())
}
```

**After running `fixrs`:**
```rust
use std::io::{Error, ErrorKind};
use std::time::Instant;

fn handle_data() -> Result<(), Error> {
    let now = Instant::now();
    let code = ErrorKind::NotFound;
    Ok(())
}
```

### Installation
```bash
cargo install fixrs
```

### Usage
```bash
# Preview
fixrs -n

# In-place rewrite and auto rustfmt
fixrs
```

It parses your code via `syn` AST, analyzes scopes to avoid name collisions, automatically injects top-level imports, and formats files with `rustfmt`.
